p2p/conn/connection.go

```diff
@@ -506,6 +537,15 @@ FOR_LOOP:
 	}
 }
 
+func (c *MConnection) stopPongTimer() {
+	if c.pongTimer != nil {
+		if !c.pongTimer.Stop() {
+			<-c.pongTimer.C
```

Comment:
suppose we receive two pongs in a row. wont we block here on the second one ?

Reply:
Here's a diff that proves the point. All I needed to do was add a sleep to the test (without the sleep the pongTimer was just nil which is why it was passing). With the sleep, the test hangs forever

```
diff --git a/p2p/conn/connection.go b/p2p/conn/connection.go
index 40f37d29..b05a36ea 100644
--- a/p2p/conn/connection.go
+++ b/p2p/conn/connection.go
@@ -484,6 +484,7 @@ FOR_LOOP:
                                break FOR_LOOP
                        }
                case packetTypePong:
+                       fmt.Println("GOT PONG")
                        c.Logger.Debug("Receive Pong")
                        c.stopPongTimer()
                case packetTypeMsg:
@@ -538,12 +539,21 @@ FOR_LOOP:
 }
 
 func (c *MConnection) stopPongTimer() {
+       fmt.Println("STOP PONG TIMER. Is nil", c.pongTimer == nil)
        if c.pongTimer != nil {
                if !c.pongTimer.Stop() {
+                       fmt.Println("waiting on .C")
                        <-c.pongTimer.C
+                       fmt.Println("done waiting")
+               } else {
+                       fmt.Println("STOP TIMER WORKED")
                }
+               fmt.Println("draining")
                drain(c.pongTimeoutCh)
+               fmt.Println("Drained")
        }
+
+       fmt.Println("-------------")
 }
 
 type ConnectionStatus struct {
diff --git a/p2p/conn/connection_test.go b/p2p/conn/connection_test.go
index 270b4ae9..6efc6e08 100644
--- a/p2p/conn/connection_test.go
+++ b/p2p/conn/connection_test.go
@@ -173,6 +173,7 @@ func TestMConnectionMultiplePongsInTheBeginning(t *testing.T) {
        err := mconn.Start()
        require.Nil(t, err)
        defer mconn.Stop()
+       time.Sleep(time.Second * 3)
 
        // sending 3 pongs in a row (abuse)
        _, err = server.Write([]byte{packetTypePong})
```

Reply:
I don't think this diff is correct (proves anything) because if you add `time.Sleep`, `sendRoutine` will attempt to send ping and block on `flush`. And since we're using a pipe here and you're not reading the ping (see `server.Read` calls), this whole thing will block.

Reply:
that could be. but do you see what I'm referring to ? It still seems like it will deadlock if it receives multiple pongs